sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/Internals/Profiling/ProfilingSessionEventSource.cs

```diff
@@ -83,6 +125,74 @@ protected override void OnEventCommand(EventCommandEventArgs command)
         /// Gets or sets the active profiling session ID.
         /// The value will be null if there is no active session.
         /// </summary>
-        public string? SessionId { get; private set; }
+        public string? SessionId
+        {
+            [NonEvent]
+            get => _sessionId;
+
+            [NonEvent]
+            private set
+            {
+                if (_sessionId != value)
+                {
+                    _sessionId = value;
+                    _sessionIdChanged?.Invoke(this, value);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Occurs when the session ID has changed, meaning either a session has started or stopped.
+        /// </summary>
+        /// <remarks>
+        /// If you subscribe to this event when a session has already started, then your handler
+        /// will be called immediately.
+        /// </remarks>
+        public event EventHandler<string?> SessionIdChanged
+        {
+            [NonEvent]
+            add
+            {
+                _sessionIdChanged += value;
+
+                // New subscribers are notified of an active session right away.
+                if (SessionId != null)
+                {
+                    value.Invoke(this, SessionId);
+                }
+            }
+
+            [NonEvent]
+            remove
+            {
+                _sessionIdChanged -= value;
+            }
+        }
+
+        /// <summary>
+        /// Write the attributes of a <see cref="Resource"/>.
+        /// </summary>
+        /// <param name="resource">The resource.</param>
+        /// <exception cref="ArgumentNullException"><paramref name="resource"/> is null.</exception>
+        [NonEvent]
+        public void WriteResourceAttributes(Resource resource)
+        {
+            if (resource is null)
+            {
+                throw new ArgumentNullException(nameof(resource));
+            }
+
+            if (IsEnabled(EventLevel.Informational, Keywords.ResourceAttributes))
+            {
+                IEnumerable<KeyValuePair<string, object>> attributes = resource.Attributes;
+                ResourceAttributes(attributes.Select(kvp => new KeyValuePair<string, string>(kvp.Key, Convert.ToString(kvp.Value, InvariantCulture)!)));
```

Comment:
`Convert.ToString(kvp.Value, InvariantCulture)` might throw an exception if customer overrides tostring and that throws.

Reply:
That is a fair point, but note that the ResourceBuilder will not let you add a non-primitive type as an attribute value ("Attribute value type is not an accepted primitive")

Reply:
Now that I look closer, it looks like resource attribute values can be arrays. I presume there's some serialization code for this that I can copy.

Reply:
I added a try/catch just in case -- although, I can't test it because the ResourceBuilder won't let me add a non-primitive type as an attribute value.

I also added length truncation.

I did not do anything for arrays. I couldn't find any existing code in the exporter that handles array attribute values. I'm thinking they're rare enough that it's OK to leave it as-is. For the profiler, the attribute values we care about (e.g. servicename) are all strings.

Reply:
would it make sense to only capture specific keys here?

We also have a precedent for checking the type of the value.
https://github.com/Azure/azure-sdk-for-net/blob/f65fb9760cbfd8894d40b3f02419d363690eb3bc/sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/ActivityTagsProcessor.cs#L87-L134

Reply:
Yeah, the attributes I'm using today are service.namespace and service.name. However, it's early days and there may be a need for others. It's easier to modify the profiler to pick out future values than it is to pre-select them here. The latter leaves us dependent on getting apps to upgrade to newer distro versions.

I didn't understand the relevance of the CategorizeTags code, @TimothyMothra. Was it to show how to string-ify arrays or a way to pick out certain kinds of attributes?

Reply:
> I didn't understand the relevance of the CategorizeTags code, @TimothyMothra. Was it to show how to string-ify arrays or a way to pick out certain kinds of attributes?

Both :)
But if you're comfortable with your current approach then I think it's fine to leave it as is.